pkg/controller/compute/v1beta1/cluster.go

```diff
@@ -0,0 +1,287 @@
+// controller and its corresponding reconciler to the manager with any runtime configuration.
+type GKEClusterController struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *GKEClusterController) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1beta1.GKEClusterGroupVersionKind),
+		resource.WithExternalConnecter(&clusterConnector{kube: mgr.GetClient(), newServiceFn: container.NewService}),
+		resource.WithManagedInitializers())
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1beta1.GKEClusterKindAPIVersion, v1beta1.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1beta1.GKECluster{}).
+		Complete(r)
+}
+
+type clusterConnector struct {
+	kube         client.Client
+	newServiceFn func(ctx context.Context, opts ...option.ClientOption) (*container.Service, error)
+}
+
+func (c *clusterConnector) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	i, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return nil, errors.New(errNotCluster)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.kube.Get(ctx, meta.NamespacedNameOf(i.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	gke, err := c.newServiceFn(ctx,
+		option.WithCredentialsJSON(s.Data[p.Spec.Secret.Key]),
+		option.WithScopes(container.CloudPlatformScope))
+	return &clusterExternal{cluster: *gke, projectID: p.Spec.ProjectID, kube: c.kube}, errors.Wrap(err, errNewClient)
+}
+
+type clusterExternal struct {
+	kube      client.Client
+	cluster   container.Service
+	projectID string
+}
+
+func (e *clusterExternal) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotCluster)
+	}
+
+	existing, err := e.cluster.Projects.Locations.Clusters.Get(meta.GetExternalName(mg)).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetCluster)
+	}
+
+	cr.Status.AtProvider = gke.GenerateObservation(*existing)
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	gke.LateInitializeSpec(&cr.Spec.ForProvider, *existing)
+	if !reflect.DeepEqual(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return resource.ExternalObservation{}, errors.Wrap(err, errManagedUpdateFailed)
+		}
+	}
+
+	// If boostrap node pool is present then we want to delete it.
+	for _, pool := range existing.NodePools {
```

Comment:
Check if 0 node "bootstrap" pool exists. If so we delete it. Alternatively, we could just try to delete it on every observation and ignore "not found" errors, but seems better to perform local operations rather than have additional network call in every reconciliation loop.